src/compiler/checker.ts

```diff
@@ -16719,12 +16715,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     // reduced we'll never reduce again, so this occurs at most once.
                     result = getIntersectionType(typeSet, aliasSymbol, aliasTypeArguments);
                 }
-                else if (eachIsUnionContaining(typeSet, TypeFlags.Undefined)) {
+                else if (every(typeSet, t => !!(t.flags & TypeFlags.Union && (t as UnionType).types[0].flags & TypeFlags.Undefined))) {
                     const containedUndefinedType = some(typeSet, containsMissingType) ? missingType : undefinedType;
                     removeFromEach(typeSet, TypeFlags.Undefined);
                     result = getUnionType([getIntersectionType(typeSet), containedUndefinedType], UnionReduction.Literal, aliasSymbol, aliasTypeArguments);
                 }
-                else if (eachIsUnionContaining(typeSet, TypeFlags.Null)) {
+                else if (every(typeSet, t => !!(t.flags & TypeFlags.Union && ((t as UnionType).types[0].flags & TypeFlags.Null || (t as UnionType).types[1].flags & TypeFlags.Null)))) {

```

Comment:
(`lines:-5+0`)

Why does this **only** need to check?
 - `types[0]` for `TypeFlags.Undefined`
 - `types[0] | types[1]` for `TypeFlags.Null`

Is there some rule/convention that `null` is first in `UnionType#types` if it's part of the union, _unless_ `undefined` is in the union (which is _really_ always first if it is part of the union)?

Reply:
Yes, union types are sorted like this internally - so if a union contains `undefined` it has to be the first element etc.